pkg/template/processor_test.go

```diff
@@ -0,0 +1,223 @@
+			"PROJECT_NAME": "foo",
+			"COMMIT":       "1a2b3c",
+			"USER_NAME":    "developer",
+			random:         random,
+		}
+		cl := test.NewFakeClient(t)
+		p := template.NewProcessor(cl, scheme, logger(t))
+
+		// when
+		objs, err := p.Process(templateContent(t), values)
+
+		// then
+		require.EqualError(t, err, fmt.Sprintf("unknown parameter name \"%s\"", random))
+		require.Len(t, objs, 0)
+	})
+}
+
+func TestProcessAndApply(t *testing.T) {
+	t.Run("ok", func(t *testing.T) {
+		s := scheme.Scheme
+		err := apis.AddToScheme(s)
+		require.NoError(t, err)
+		utilruntime.Must(apitemplate.Install(s)) // see https://github.com/openshift/oc/blob/master/cmd/oc/oc.go#L77
+
+		templateContent := templateContent(t)
+		pn := uuid.NewV4().String()
+		u := uuid.NewV4().String()
+		c := uuid.NewV4().String()
+		values := map[string]string{
+			"PROJECT_NAME": pn,
+			"COMMIT":       c,
+			"USER_NAME":    u,
+		}
+
+		cl := test.NewFakeClient(t)
+		p := template.NewProcessor(cl, s, logger(t))
+
+		err = p.ProcessAndApply(templateContent, values)
+		require.NoError(t, err)
+		// check that the project request was created
+		verifyProjectRequest(t, cl, pn)
+		verifyRoleBinding(t, cl, pn)
+	})
+
+	t.Run("delete role binding and apply template", func(t *testing.T) {
+		// given
+		s := scheme.Scheme
+		err := apis.AddToScheme(s)
+		require.NoError(t, err)
+		utilruntime.Must(apitemplate.Install(s)) // see https://github.com/openshift/oc/blob/master/cmd/oc/oc.go#L77
+
+		templateContent := templateContent(t)
+		pn := uuid.NewV4().String()
+		u := uuid.NewV4().String()
+		c := uuid.NewV4().String()
+		values := map[string]string{
+			"PROJECT_NAME": pn,
+			"COMMIT":       c,
+			"USER_NAME":    u,
+		}
+
+		cl := test.NewFakeClient(t)
+		p := template.NewProcessor(cl, s, logger(t))
+
+		err = p.ProcessAndApply(templateContent, values)
+		require.NoError(t, err)
+
+		verifyProjectRequest(t, cl, pn)
+		verifyRoleBinding(t, cl, pn)
+
+		// delete rolebinding to create scenario, of rolebinding failed to create in first run.
+		rb, err := roleBinding(cl, pn)
+		require.NoError(t, err)
+
+		err = cl.Delete(context.TODO(), rb)
+		require.NoError(t, err)
+
+		// apply the same templates
+		err = p.ProcessAndApply(templateContent, values)
+		require.NoError(t, err)
+
+		verifyProjectRequest(t, cl, pn)
+		verifyRoleBinding(t, cl, pn)
+	})
+}
+
+
+func templateContent(t *testing.T) []byte {
+	templateContent, err := GetTemplateContent("basic-tier-template.yml")
+	require.NoError(t, err)
+
+	return templateContent
+}
+
+func logger(t *testing.T) logr.Logger {
+	zapLog, err := zap.NewDevelopment()
+	require.NoError(t, err)
+	return zapr.NewLogger(zapLog)
+}
+
+func expectedRoleBinding() string {
+	return `{"apiVersion":"authorization.openshift.io/v1","kind":"RoleBinding","metadata":{"labels":{"provider":"codeready-toolchain","version":"1a2b3c"},"name":"foo-admin","namespace":"foo"},"roleRef":{"name":"admin"},"subjects":[{"kind":"User","name":"developer"}]}
```

Comment:
I have already written one related comment, having the template in a file and then expected processed outcome as a string is not the best

Reply:
hmm, let me think how can we change this to make it maintainable in less effort.

Reply:
take a look at tenant service tests - There are the same tests